doc/ts_config.3
.\" Copyright (c) 2017, Martin Kepplinger <martink@posteo.de>
.\"
.\" %%%LICENSE_START(GPLv2+_DOC_FULL)
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, see
.\" <http://www.gnu.org/licenses/>.
.\" %%%LICENSE_END
.\"
.TH TS_CONFIG 3  "" "" "tslib"
.SH NAME
ts_config, ts_reconfig \- read tslib's configuration
.SH SYNOPSIS
.nf
.B #include <tslib.h>
.sp
.BI "int ts_config(struct tsdev *" dev ");"
.sp
.BI "int ts_reconfig(struct tsdev *" dev ");"
.sp
.fi

.SH DESCRIPTION
.BR ts_config ()
loads and initialises all modules configured in the tslib configuration file
.BR TSLIB_CONFFILE
, see ts.conf (5).
.BR ts_reconfig ()
re-runs ts_config and updates the configuration if the configuration file has changed.

.SH RETURN VALUE
Zero is returned on success. A negative value is returned in case of an error.

.SH SEE ALSO
.BR ts_read (3),
.BR ts_open (3),
.BR ts_setup (3),
.BR ts_close (3),
.BR ts.conf (5)


NEWS
tslib 1.11 - tbd
================================

This release includes libts version 0.7.1 and the following changes:
* The graphical tools don't show TSLIB anymore, but Touchscreen instead.
* The graphical tools support rotating the screen 90, 180 and 270 degrees
* man page updates

This release includes the following bugfixes:
* ff8c7d9 use dlfcn.h dynamic loading function only if libdl is available
* d39d358 input-raw: fix multitouch devices without BTN_TOUCH


tslib 1.10 - released 2017-05-12
================================

This release includes libts version 0.7.0 and the following changes:
* new `ts_libversion()` API and TSLIB_VERSION_MT definition
  This offers seamless integration for apps that don't know tslib's version
* new `module_raw dmc_dus3000` for DMC DUS Series touch screen via UART
* Fixed and improved statically linked building

This release includes the following bugfixes:
* c03f7c9 tools: ts_uinput_start.sh: fix usage of ts_uinput -d -v
* 47e7ab5 input-raw: Fix wrong first slot values by using calloc instead of malloc
* 7c5b4f8 plugins: filters: check for read_mt() availability
* e684df5 call ops->fini() only when implemented by the plugin
* 0ca9580 ts_close: free tslib_module_info when fini() is not implemented
* 5974266 median: read_mt: mark all-zero samples invalid


tslib 1.9 - released 2017-04-21
===============================

This release includes libts version 0.6.0 and the following changes:
* new `module iir` infinite impulse response filter
* `module_raw galax` is deprecated. Users should switch to `module_raw input`
* multitouch for (at least) eGalax 210 (`module_raw input` and `ts_read_mt()`)
* new `ts_verify` API test program
* `ts_uinput -d -v` prints the new /dev/input/eventX device node before returning
* experimental support for GNU / Hurd

This release includes the following bugfixes:
* fix support for protocol type A devices (for example Sitronix ST1232)
* fix the `skip` filter's read_mt() multitouch operation
* 338d2e4 input-raw: add MT protocol type A support to ts_read()
* e47df79 input-raw: fix pressure capable multitouch devices
* dbb8f4c ts_uinput: correctly return in case forking fails
* 73773ac fbutils-linux: fix framebuffer errors on 64 bit systems
* 4d63bfa input-raw: fix ts_read_mt for non-MT devices without pressure


tslib 1.8 - released 2017-03-21
===============================

tslib 1.8 is backwards compatible. Changes since last release include:
* experimental support for building on Windows (using MSYS2)
* fixed ts_read_mt() in non-blocking operation when trying to read more than
  1 sample, see bugfixes
* added workaround for strangely behaving eGalax model 210 devices, see bugfixes
* bugfixes
  - 388d3da input-raw: ts_read_mt: fix returning actually read samples
  - c7f0966 debounce: use signed time values
  - 9a9b0c9 pthres: read_mt: fix internal pressure state handling
  - fa10e4c galax: version 210: use ABS_MT_DISTANCE to find pressure 0


tslib 1.7 - released 2017-03-13
===============================

Changes since last release:
* ts_read_mt() now consistently returns errno set by read() when using
  module_raw input
* extended ts_print_mt to test more of our API
* ts_print_mt --raw replaces the ts_print_raw_mt program
* basic support for the Haiku OS
* added a release helper script to simplify the releasing process
* add proper AUTHORS and THANKS files, as well as .mailmap
* bugfixes
  - 6ea643e ts_test_mt: fix idev commandline option
  - 9617c7d dmc-raw: fix errno return code to be negative
  - dc2fdaf input-raw: remove errror message on each negative return from read()
  - 75b5f40 input-raw: return read() errno on error.
  - de5e261 skip: fix buffer memory freeing
  - da5477e median: fix ts_read_mt() pen down event


tslib 1.6 - released 2017-03-03
===============================

Changes since the last release:
* ABI addition: libts.so.0 (ensuring backwards compatibility)
* tools/ts_uinput_start.sh that creates /dev/input/ts_uinput
* improved documentation in README
* support for FreeBSD (experimental)
* bugfixes and performance improvements
  - 3b01897 skip: fix order of memory freeing in finish routine
  - 9ed4eca skip: avoid unnecessary malloc during operation
  - f9202cc median: fix memory cleanup
  - 69444b9 input-raw: improve input definitions backwards compatibility


tslib 1.5 - released 2017-02-09
===============================

Changes since 1.4 are:
* restore full backwards compatibility for all old kernel headers
* 64 bit compatibility for test programs
* bugfixes
  - 3f99a83 Revert "plugins: linear: fix config file access"


tslib 1.4 - released 2017-02-03
===============================

Most notable changes since 1.3 are:
* added default input device names (for example /dev/input/ts)
* new API: ts_setup()
* support for Android (experimental)
* man page updates and fixes
* compatibility fixes
* bugfixes all over the place


tslib 1.3 - released 2017-01-09
===============================

Most notable changes since 1.2 are:
* user visible:
    - multitouch:
      tests/ts_print_mt
      tests/ts_print_raw_mt
      tests/ts_test_mt
    - tools/ts_uinput: daemon program offers an input event device to use, instead of tslib's API
    - new filter plugin: median filter
    - man pages

* user API:
    - ts_read_mt() and struct ts_sample_mt
    - ts_reconfig()

* misc:
    - support for multitouch in input-raw and filter plugins
    - bugfixes (partly found by Coverity)
    - build system fixes


tslib 1.2 - released 2016-11-18
===============================

Most notable changes since 1.1 are:
* user visible:
    new plugins, see below
    added ts_finddev program (via buildroot)

* hardware plugins:
    improved eGalax touchscreen support
    added support for eGalax 7200 and 7201 touchscreens
    added hardware plugin for WaveShare touchscreens

* filter plugins:
    bugfixes in variance plugin
    added debounce plugin
    added skip plugin

* misc:
    library versioning

src/ts_config.c
/*
 *  tslib/src/ts_config.c
 *
 *  Copyright (C) 2001 Russell King.
 *
 * This file is placed under the LGPL.  Please see the file
 * COPYING for more details.
 *
 *
 * Read the configuration and load the appropriate drivers.
 */
#include "config.h"
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#ifdef HAVE_LIBDL
#include <dlfcn.h>
#endif

#include <errno.h>

#if !defined(HAVE_STRSEP)
#include "ts_strsep.h"
#endif

#include "tslib-private.h"

/* Maximum line size is BUF_SIZE - 2
 * -1 for fgets and -1 to test end of line
 */
#define BUF_SIZE 512

/* Discard any empty tokens
 * Note: strsep modifies p (see man strsep)
 */
static void discard_null_tokens(char **p, char **tokPtr)
{
	while (*p != NULL && **tokPtr == '\0') {
	#if !defined HAVE_STRSEP
		*tokPtr = ts_strsep(p, "\t");
	#else
		*tokPtr = strsep(p, " \t");
	#endif
	}
}

int ts_config(struct tsdev *ts)
{
	char buf[BUF_SIZE], *p;
	FILE *f;
	int line = 0;
	int ret = 0;
	short strdup_allocated = 0;

	char *conffile;

	if ((conffile = getenv("TSLIB_CONFFILE")) == NULL) {
		conffile = strdup(TS_CONF);
		if (conffile) {
			strdup_allocated = 1;
		} else {
			ts_error("Couldn't find tslib config file: %s\n",
				 strerror(errno));
			return -1;
		}
	}

	f = fopen(conffile, "r");
	if (!f) {
		if (strdup_allocated)
			free(conffile);

		ts_error("Couldn't open tslib config file: %s\n",
			 strerror(errno));
		return -1;
	}

	buf[BUF_SIZE - 2] = '\0';
	while ((p = fgets(buf, BUF_SIZE, f)) != NULL) {
		char *e;
		char *tok;
		char *module_name;

		line++;

		/* Chomp */
		e = strchr(p, '\n');
		if (e)
			*e = '\0';

		/* Did we read a whole line? */
		if (buf[BUF_SIZE - 2] != '\0') {
			ts_error("%s: line %d too long\n", conffile, line);
			break;
		}

	#if !defined HAVE_STRSEP
		tok = ts_strsep(&p, " \t");
	#else
		tok = strsep(&p, " \t");
	#endif
		discard_null_tokens(&p, &tok);

		/* Ignore comments or blank lines.
		 * Note: strsep modifies p (see man strsep)
		 */
		if (p == NULL || *tok == '#')
			continue;

		/* Search for the option. */
		if (strcasecmp(tok, "module") == 0) {
		#if !defined HAVE_STRSEP
			module_name = ts_strsep(&p, " \t");
		#else
			module_name = strsep(&p, " \t");
		#endif
			discard_null_tokens(&p, &module_name);
			ret = ts_load_module(ts, module_name, p);
		} else if (strcasecmp(tok, "module_raw") == 0) {
		#if !defined HAVE_STRSEP
			module_name = ts_strsep(&p, " \t");
		#else
			module_name = strsep(&p, " \t");
		#endif
			discard_null_tokens(&p, &module_name);
			ret = ts_load_module_raw(ts, module_name, p);
		} else {
			ts_error("%s: Unrecognised option %s:%d:%s\n",
				 conffile, line, tok);
			break;
		}
		if (ret != 0) {
			ts_error("Couldnt load module %s\n", module_name);
			break;
		}
	}

	if (ts->list_raw == NULL) {
		ts_error("No raw modules loaded.\n");
		ret = -1;
	}

	fclose(f);

	if (strdup_allocated)
		free(conffile);

	return ret;
}

int ts_reconfig(struct tsdev *ts)
{
	void *handle;
	int ret;
	struct tslib_module_info *info, *next;
	int fd;

	info = ts->list;
	while (info) {
		/* Save the "next" pointer now because info will be freed */
		next = info->next;

		handle = info->handle;

		if (info->ops->fini)
			info->ops->fini(info);
	#ifdef HAVE_LIBDL
		if (handle)
			dlclose(handle);
	#endif

		info = next;
	}

	fd = ts->fd;	/* save temp */
	memset(ts, 0, sizeof(struct tsdev));
	ts->fd = fd;

	ret = ts_config(ts);
	return ret;
}
